SONARJAVA-5428 Create rule S7465: Unused arguments in lambda expressions should be denoted with an underscore "_" (#4857)
SonarSource/rspec: rules/S7465/java/rule.adoc

== Why is this an issue?

> Library authors often design APIs to be versatile, enabling them to handle diverse scenarios. Consequently, when we provide a lambda expression as an argument to a library function, we may find ourselves with unused arguments. Naming these arguments can contribute to unnecessary code expansion and diminish readability. Furthermore, it obscures the crucial fact that the argument is intentionally unused. To resolve this and clearly indicate the intent, unused arguments should be denoted with an underscore `_`, a feature introduced in Java 22.

Reviewer: For readability and easing future maintenance, we would benefit from splitting each sentence on a separate line

Reviewer 2: Done. IIRC, in the past I've been asked to do the opposite, but I like separate lines too.

> Library authors often design APIs to be versatile, enabling them to handle diverse scenarios. Consequently, when we provide a lambda expression as an argument to a library function, we may find ourselves with unused arguments. Naming these arguments can contribute to unnecessary code expansion and diminish readability. Furthermore, it obscures the crucial fact that the argument is intentionally unused. To resolve this and clearly indicate the intent, unused arguments should be denoted with an underscore `_`, a feature introduced in Java 22.

Reviewer: What does `code expansion` mean in this context? My first google search leads me to genetic topics

Reviewer: `it obscures the crucial fact`, Isn't that a little too dramatic?

Reviewer: What about moving the last sentence to the `How to fix it` section?

Reviewer 2: Done. Rewritten.

> Library authors often design APIs to be versatile, enabling them to handle diverse scenarios. Consequently, when we provide a lambda expression as an argument to a library function, we may find ourselves with unused arguments. Naming these arguments can contribute to unnecessary code expansion and diminish readability. Furthermore, it obscures the crucial fact that the argument is intentionally unused. To resolve this and clearly indicate the intent, unused arguments should be denoted with an underscore `_`, a feature introduced in Java 22.

Reviewer: `... enabling them to handle diverse scenarios. Consequently, when we provide a lambda expression ...`
I see how it connects in context, but the jump from a library author handling diverse situations and US providing a lambda could use a bridge

> Library authors often design APIs to be versatile, enabling them to handle diverse scenarios. Consequently, when we provide a lambda expression as an argument to a library function, we may find ourselves with unused arguments. Naming these arguments can contribute to unnecessary code expansion and diminish readability. Furthermore, it obscures the crucial fact that the argument is intentionally unused. To resolve this and clearly indicate the intent, unused arguments should be denoted with an underscore `_`, a feature introduced in Java 22.

Reviewer: ```
To resolve this and clearly indicate the intent, unused arguments should be denoted with an underscore `_`, a feature introduced in Java 22.
```
Can we use a more active?
```
Java 22 introduces the unnamed variable pattern `_` that clearly indicates the intent not to use the variable.
```

Reviewer 2: Done.

== How to fix it

=== Code examples

==== Noncompliant code example
[source,java,diff-id=1,diff-type=noncompliant]
----

> BiFunction<Integer,Integer,Integer> f = (a, b) -> a; // Noncompliant
> 
> // ...
> 
> firstValueMap.merge(event.id, event.value, (oldValue, newValue) -> oldValue);  // Noncompliant

Reviewer: Let's make the unused variables more obvious
```suggestion
BiFunction<Integer,Integer,Integer> f = (a, unused) -> a; // Noncompliant

// ...

firstValueMap.merge(event.id, event.value, (oldValue, unused) -> oldValue);  // Noncompliant
```

----

==== Compliant solution
[source,java,diff-id=1,diff-type=compliant]
----
BiFunction<Integer,Integer,Integer> f2 = (a, _) -> a;

// ...

firstValueMap.merge(event.id, event.value, (oldValue, _) -> oldValue);
----

== Resources
=== Documentation

> * Java Language Updates - https://docs.oracle.com/en/java/javase/22/language/unnamed-variables-and-patterns.html[Unnamed Variables and Patterns]

Reviewer: I would personnally prefer a link to the JEP itself?
```suggestion
* OpenJDK - https://openjdk.org/jeps/456[JEP 456: Unnamed Variables & Patterns]
```

== Why is this an issue?

Library authors often design versatile APIs to handle diverse scenarios.

> When such an API takes a function as a parameter, developers may find themselves passing a lambda expression where some parameters are unused.
> Naming these parameters can lead to unnecessary verbosity.
> Java 22 introduces the unnamed variable pattern `_` that clearly indicates the intent not to use the variable.

Reviewer: ```suggestion
When such an API takes a function as a parameter, developers may find themselves passing a lambda expression where some parameters are unused.
Naming such parameters is unnecessary and creates distraction.
Java 22 introduces the unnamed variable pattern `_` that clearly indicates the intent not to use them.
```

== How to fix it
In order to clearly indicate the intent, unused parameters should be denoted with an unnamed variable pattern `_`.

=== Code examples

==== Noncompliant code example
[source,java,diff-id=1,diff-type=noncompliant]
----
BiFunction<Integer, Integer, Integer> f = (a, unused) -> a; // Noncompliant
BiFunction<Integer, Integer, Integer> g = (unused1, unused2) -> 64; // Noncompliant

// ...
Map<String, Integer> firstValueMap = // ...
firstValueMap.merge(event.id, event.value, (oldValue, unused) -> oldValue);  // Noncompliant
----

==== Compliant solution
[source,java,diff-id=1,diff-type=compliant]
----
BiFunction<Integer, Integer, Integer> f2 = (a, _) -> a;
BiFunction<Integer, Integer, Integer> g2 = (_, _) -> 64;

// ...
Map<String, Integer> firstValueMap = // ...

> firstValueMap.merge(event.id, event.value, (oldValue, _) -> oldValue);

Reviewer: Let's wrap this in a function

== Why is this an issue?

Library authors often design versatile APIs to handle diverse scenarios.

> When such an API takes a function as an argument, we may find ourselves
> writing a lambda expression where some arguments are unused.

Reviewer: How about removing ourselves from the description?
```suggestion
When such an API takes a function as an argument, developers may find themselves passing a lambda expression where some arguments are unused.
```

Naming these arguments can lead to unnecessary verbosity.
Java 22 introduces the unnamed variable pattern `_` that clearly indicates
the intent not to use the variable.

== How to fix it
In order to clearly indicate the intent, unused arguments should be denoted
with an unnamed variable pattern `_`.

=== Code examples

==== Noncompliant code example
[source,java,diff-id=1,diff-type=noncompliant]
----

> BiFunction<Integer,Integer,Integer> f = (a, unused) -> a; // Noncompliant

Reviewer: The description above mentions that **some** arguments may be unused, what about adding an example where multiple arguments are unused?
